File: 백준/Gold/28118. 안전한 건설 계획/안전한 건설 계획.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;
import java.util.stream.IntStream;


public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[] parent = IntStream.range(0, n+1).toArray();
        int[] rank = new int[n+1];
        Arrays.fill(rank, 1);

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            union(parent, rank, u, v);
        }

        int count = 0;
        for (int i = 1; i < n; i++) {
            if (find(parent, i) != find(parent, i+1)) {
                union(parent, rank, i, i+1);
                count++;
            }
        }
        System.out.println(count);
    }

    public static int find(int[] parent, int x) {
        if (x == parent[x])
            return x;
        parent[x] = find(parent, parent[x]);
        return parent[x];
    }


    public static void union(int[] parent, int[] rank, int x, int y) {
        int px = find(parent, x);
        int py = find(parent, y);

        if (px == py)
            return;

        if (rank[px] >= rank[py]) {
            parent[py] = px;
            rank[px] += rank[py];
        } else {
            parent[px] = py;
            rank[py] += rank[px];
        }
    }
}

File: 백준/Gold/28118. 안전한 건설 계획/안전한 건설 계획.py
# https://www.acmicpc.net/problem/28118
import sys
input = sys.stdin.readline


def find(x):
    if x == parent[x]:
        return x
    parent[x] = find(parent[x])
    return parent[x]


def union(x, y):
    x, y = map(find, (x, y))

    if x == y:
        return

    if rank[x] >= rank[y]:
        parent[y] = x
        rank[x] += rank[y]
    else:
        parent[x] = y
        rank[y] += rank[x]


n, m = map(int, input().split())
parent = list(range(n+1))
rank = [1] * (n+1)
for _ in range(m):
    union(*map(int, input().split()))

cnt = 0
for i in range(1, n):
    if find(i) != find(i+1):
        union(i, i+1)
        cnt += 1

print(cnt)